keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java

```diff
@@ -0,0 +1,137 @@
+
+        assertEquals(0., KLDivergence.compute(qDist, samples), 1e-6);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void throwsExceptionIfPProbIsNotZeroButQIs() {
+        QDistribution q = mock(QDistribution.class);
+        when(q.getLogOfMasterP(any(NetworkState.class))).thenReturn(Double.NEGATIVE_INFINITY);
+        NetworkSamples samples = createNetworkSamplesWithOneVertexAndOneSample(Math.log(0.5));
+
+        KLDivergence.compute(q, samples);
+    }
+
+    @Test
+    public void klDivergenceReturnsZeroIfPIsZeroButQIsNot() {
+        QDistribution q = mock(QDistribution.class);
+        when(q.getLogOfMasterP(any(NetworkState.class))).thenReturn(Math.log(0.5));
+        NetworkSamples samples = createNetworkSamplesWithOneVertexAndOneSample(Double.NEGATIVE_INFINITY);
+
+        assertEquals(0., KLDivergence.compute(q, samples), 1e-10);
+    }
+
+    @Test
+    public void klDivergenceReturnsZeroIfPAndQAreZero() {
+        QDistribution q = mock(QDistribution.class);
+        when(q.getLogOfMasterP(any(NetworkState.class))).thenReturn(Double.NEGATIVE_INFINITY);
+        NetworkSamples samples = createNetworkSamplesWithOneVertexAndOneSample(Double.NEGATIVE_INFINITY);
+
+        assertEquals(0., KLDivergence.compute(q, samples), 1e-10);
+    }
+
+    @Test
+    public void returnsLargerKLDivergenceIfTheLocationOfQIsFurtherFromP_QIsProbabilisticDouble() {
+        GaussianVertex v1 = new GaussianVertex(0., 1.);
+        ConstantDoubleVertex v2 = new ConstantDoubleVertex(0.1);
+        DoubleVertex v3 = v1.plus(v2);
+
+        BayesianNetwork network = new BayesianNetwork(v3.getConnectedGraph());
+        NetworkSamples samples = MetropolisHastings
+            .withDefaultConfig()
+            .getPosteriorSamples(network, Collections.singletonList(v1), 1000);
+
+        ProbabilisticDouble q1 = new GaussianVertex(0.1, 1.);
+        ProbabilisticDouble q2 = new GaussianVertex(10.0, 1.);
+
+        assertThat(KLDivergence.compute(q1, samples), lessThan(KLDivergence.compute(q2, samples)));
+    }
+
+    @Test
+    public void returnsLargerKLDivergenceIfTheLocationOfQIsFurtherFromP_QIsQDistribution() {
+        GaussianVertex v1 = new GaussianVertex(0., 1.);
+        ConstantDoubleVertex v2 = new ConstantDoubleVertex(0.1);
+        DoubleVertex v3 = v1.plus(v2);
+
+        BayesianNetwork network = new BayesianNetwork(v3.getConnectedGraph());
+        NetworkSamples samples = MetropolisHastings
+            .withDefaultConfig()
+            .getPosteriorSamples(network, Collections.singletonList(v1), 1000);
+
+        QDistribution q1 = new TestGaussianQDistribution(0.1, 1.);
+        QDistribution q2 = new TestGaussianQDistribution(10.0, 1.);
+
+        assertThat(KLDivergence.compute(q1, samples), lessThan(KLDivergence.compute(q2, samples)));
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void throwsExceptionIfNetworkStateHasMoreThanOneVertexAndQIsProbabilisticDouble(){
```

Comment:
I think this could do with some explanation in a javadoc comment. Why is this a bad thing to do? Is this just something that's not yet supported, or is it a logical mistake on the user's part? If the latter - what should the user do instead?

Reply:
A neat library you can use here is https://junit.org/junit4/javadoc/latest/org/junit/rules/ExpectedException.html#expectMessage(java.lang.String)